cmd/juju/model/setcredential.go

```diff
@@ -0,0 +1,188 @@
+	}
+	if !names.IsValidCloudCredentialName(args[1]) {
+		return errors.NotValidf("cloud credential name %q", args[1])
+	}
+	c.cloud = args[0]
+	c.credential = args[1]
+	return nil
+}
+
+// Run implements Command.Run.
+func (c *modelCredentialCommand) Run(ctx *cmd.Context) error {
+	fail := func(e error) error {
+		ctx.Infof("Failed to change model credential: %v", e)
+		return e
+	}
+
+	root, err := c.newAPIRootFunc()
+	if err != nil {
+		return fail(errors.Annotate(err, "opening API connection"))
+	}
+	defer root.Close()
+
+	accountDetails, err := c.CurrentAccountDetails()
+	if err != nil {
+		return fail(errors.Annotate(err, "getting current account"))
+	}
+	userTag := names.NewUserTag(accountDetails.User)
+	cloudTag := names.NewCloudTag(c.cloud)
+	credentialTag, err := common.ResolveCloudCredentialTag(userTag, cloudTag, c.credential)
+	if err != nil {
+		return fail(errors.Annotate(err, "resolving credential tag"))
+	}
+
+	cloudClient := c.newCloudAPIFunc(root)
+	defer cloudClient.Close()
+
+	remote := false
+	remoteCredentials, err := cloudClient.UserCredentials(userTag, cloudTag)
+	if err != nil {
+		// This is ok - we can proceed with local ones anyway.
+		ctx.Infof("Could not determine if there are remote credentials for the user: %v", err)
+	} else {
+		for _, credTag := range remoteCredentials {
+			if credTag == credentialTag {
+				remote = true
+				ctx.Infof("Found credential remotely, on the controller. Not looking locally...")
+				break
+			}
+		}
+	}
+
+	// Credential does not exist remotely. Upload it.
+	if !remote {
+		ctx.Infof("Did not find credential remotely. Looking locally...")
+		credential, err := c.findCredentialLocally(ctx)
+		if err != nil {
+			return fail((err))
+		}
+		ctx.Infof("Uploading local credential to the controller.")
+		err = cloudClient.AddCredential(credentialTag.String(), *credential)
+		if err != nil {
+			return fail(err)
+		}
+	}
+
+	_, modelDetails, err := c.ModelDetails()
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	modelTag := names.NewModelTag(modelDetails.ModelUUID)
+
+	modelClient := c.newModelCredentialAPIFunc(root)
+	defer modelClient.Close()
+
+	err = modelClient.ChangeModelCredential(modelTag, credentialTag)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	ctx.Infof("Changed model credential.")
+	return nil
+}
+
+func (c *modelCredentialCommand) findCredentialLocally(ctx *cmd.Context) (*cloud.Credential, error) {
+	foundcloud, err := common.CloudByName(c.cloud)
+	if err != nil {
+		return nil, err
+	}
+	getCredentialsParams := modelcmd.GetCredentialsParams{
+		Cloud:          *foundcloud,
+		CredentialName: c.credential,
+	}
+	credential, _, _, err := modelcmd.GetCredentials(ctx, c.ClientStore(), getCredentialsParams)
+	if err != nil {
+		return nil, err
+	}
+	return credential, nil
+}
+
+const modelCredentialDoc = `This command changes a cloud credential used by a model.
+
+When current user has specified credential for the given cloud remotely, i.e
```

Comment:
When the current user has the specified credential for the cloud on the controller...